test/ZNSSubRegistrar.test.ts

```diff
@@ -122,22 +136,423 @@ describe("ZNSSubRegistrar", () => {
+      const subdomainObj = {
+        parentHash: rootHash,
+        label: "subdomain1",
+        domainAddress: admin.address,
+        tokenURI: "0://tokenURI",
+        distributionConfig: {
+          pricerContract: await zns.curvePricer.getAddress(),
+          paymentType: PaymentType.STAKE,
+          accessType: AccessType.LOCKED,
+        },
+        paymentConfig: {
+          token: await zns.meowToken.getAddress(),
+          beneficiary: admin.address,
+        },
+      };
+
+      // Add allowance
+      await zns.meowToken.connect(lvl2SubOwner).approve(await zns.treasury.getAddress(), ethers.MaxUint256);
+
+      await expect(
+        zns.subRegistrar.connect(lvl2SubOwner).registerMultipleSubdomains([subdomainObj, subdomainObj])
+      ).to.be.revertedWithCustomError(zns.subRegistrar, "DomainAlreadyExists");
+    });
+
+    it("Should revert with #ZeroAddressPassed when 1st level subdomain has zerohash", async () => {
+      const subdomainObj = {
+        parentHash: ethers.ZeroHash,
+        label: "subdomainzeroaAddresspassed",
+        domainAddress: admin.address,
+        tokenURI: "0://tokenURI",
+        distributionConfig: {
+          pricerContract: await zns.curvePricer.getAddress(),
+          paymentType: PaymentType.STAKE,
+          accessType: AccessType.LOCKED,
+        },
+        paymentConfig: {
+          token: await zns.meowToken.getAddress(),
+          beneficiary: admin.address,
+        },
+      };
+
+      // Add allowance
+      await zns.meowToken.connect(lvl2SubOwner).approve(await zns.treasury.getAddress(), ethers.MaxUint256);
+
+      await expect(
+        zns.subRegistrar.connect(lvl2SubOwner).registerMultipleSubdomains([subdomainObj])
+      ).to.be.revertedWithCustomError(zns.subRegistrar, ZERO_PARENTHASH_ERR);
+    });
+
+    it("Should register a mix of nested and non-nested subdomains and validates hashes", async () => {
+      // Structure of domains (- rootDomainName; + subdomain):
+      //
+      // - rootHash
+      //   + nested0
+      //    + nested1
+      //     + nested2
+      //      + nested3
+      // - specificParentHash_0
+      //   + non0nested0with0specific0parent0
+      // - specificParentHash_1
+      //   + non0nested0with0specific0parent1
+      //    + after0non0nested0
+      //     + after0non0nested1
+      // - rootHash
+      //   + non0nested0
+      //   + non0nested1
+
+      const subRegistrations = [];
+      const labels = [
+        "nested0",
+        "nested1",
+        "nested2",
+        "nested3",
+        "non0nested0with0specific0parent0",
+        "non0nested0with0specific0parent1",
+        "after0non0nested0",
+        "after0non0nested1",
+        "non0nested0",
+        "non0nested1",
+      ];
+      const specificParentHashes = [];
+      const expectedHashes : Array<string> = [];
+
+      // 4 nested: first has parentHash, the rest have parentHash = 0 (nested)
+      for (let i = 0; i < 4; i++) {
```

Comment:
You have over 140 lines here dedicated to setup for the test. Here and many other places will benefit from abstraction and creating a helper function to do this part for you. please abstract all of this and below into helpers instead of generating each one manually and uniquely. This is all pre-test configuration for what you're **actually** testing.